src/XAD/Jacobian.hpp

```diff
@@ -0,0 +1,131 @@
+******************************************************************************/
+
+#pragma once
+
+#include <XAD/TypeTraits.hpp>
+#include <XAD/XAD.hpp>
+
+#include <functional>
+#include <type_traits>
+#include <vector>
+
+namespace xad
+{
+// adj 2d vector
+template <typename T>
+std::vector<std::vector<T>> computeJacobian(
+    const std::vector<xad::AReal<T>> &vec,
+    std::function<std::vector<xad::AReal<T>>(std::vector<xad::AReal<T>> &)> foo,
+    xad::Tape<T> *tape = xad::Tape<T>::getActive())
+{
+    auto v(vec);
+    std::vector<std::vector<T>> matrix(foo(v).size(), std::vector<T>(v.size(), 0.0));
+    computeJacobian(vec, foo, begin(matrix), end(matrix), tape);
+    return matrix;
+}
+
+// adj iterator
+template <class RowIterator, typename T>
+void computeJacobian(const std::vector<xad::AReal<T>> &vec,
+                     std::function<std::vector<xad::AReal<T>>(std::vector<xad::AReal<T>> &)> foo,
+                     RowIterator first, RowIterator last,
+                     xad::Tape<T> *tape = xad::Tape<T>::getActive())
+{
+    auto v(vec);
+    unsigned int domain = static_cast<unsigned int>(vec.size()),
+                 codomain = static_cast<unsigned int>(foo(v).size());
+
+    if (std::distance(first, last) != codomain)
+        throw OutOfRange("Iterator not allocated enough space");
+    static_assert(
+        xad::detail::has_begin<typename std::iterator_traits<RowIterator>::value_type>::value,
+        "RowIterator must dereference to a type that implements a begin() method");
+    std::unique_ptr<xad::Tape<T>> t;
+    if (!tape)
+    {
+        t = std::unique_ptr<xad::Tape<T>>(new xad::Tape<T>());
+        tape = t.get();
+    }
+
+    tape->registerInputs(v);
+
+    auto row = first;
+    for (unsigned int i = 0; i < codomain; i++, row++)
+    {
+        auto col = row->begin();
+        for (unsigned int j = 0; j < domain; j++, col++)
+        {
+            tape->newRecording();
+            xad::AReal<T> y = foo(v)[i];
+            tape->registerOutput(y);
+            derivative(y) = 1.0;
+            tape->computeAdjoints();
+            *col = derivative(v[j]);
+        }
+    }
+}
+
+// fwd 2d vector
+template <typename T>
+std::vector<std::vector<T>> computeJacobian(
+    const std::vector<xad::FReal<T>> &vec,
+    std::function<std::vector<xad::FReal<T>>(std::vector<xad::FReal<T>> &)> foo)
+{
+    auto v(vec);
+    std::vector<std::vector<T>> matrix(foo(v).size(), std::vector<T>(v.size(), 0.0));
+    computeJacobian(vec, foo, begin(matrix), end(matrix));
+    return matrix;
+}
+
+// fwd iterator
+template <class RowIterator, typename T>
+void computeJacobian(const std::vector<xad::FReal<T>> &vec,
+                     std::function<std::vector<xad::FReal<T>>(std::vector<xad::FReal<T>> &)> foo,
+                     RowIterator first, RowIterator last)
+{
+    auto v(vec);
+    unsigned int domain = static_cast<unsigned int>(vec.size()),
+                 codomain = static_cast<unsigned int>(foo(v).size());
```

Comment:
Same here - you're executing the function just to get the output size. That's a useless execution.

Reply:
I'm not sure how I can compute the codomain without evaluating the function. I'll resolve the other conversations because they're essentially all about this issue. I've made some changes that make a slight difference but I don't think it gets any more efficient than that without a counterintuitive overhead or a bloat of new arguments.